File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/item/ItemService.java
package top.likoslupus.cellulosesz.api.item;

import top.likoslupus.cellulosesz.api.platform.CellPlayer;
import top.likoslupus.cellulosesz.api.platform.operation.PlatformResult;

import java.util.Set;

/** Item application service that preserves validation and platform failure states. */
public interface ItemService {

    Set<String> itemNames();

    PlatformResult<ItemDescriptor> parse(String input);

    PlatformResult<Boolean> valid(ItemDescriptor item);

    boolean blacklisted(ItemDescriptor item);

    PlatformResult<Integer> maxStackSize(ItemDescriptor item);

    PlatformResult<ItemGrantResult> give(CellPlayer player, ItemDescriptor item);

    PlatformResult<Integer> count(CellPlayer player, ItemDescriptor item);

    PlatformResult<Void> take(CellPlayer player, ItemDescriptor item);

    PlatformResult<String> heldItemId(CellPlayer player);

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/item/ItemService.kt
package top.likoslupus.cellulosesz.api.item

import top.likoslupus.cellulosesz.api.platform.CellPlayer
import top.likoslupus.cellulosesz.api.platform.operation.PlatformResult

/** Item application service that preserves validation and platform failure states. */
public interface ItemService {

    public fun itemNames(): Set<String>

    public fun parse(input: String): PlatformResult<ItemDescriptor>

    public fun valid(item: ItemDescriptor): PlatformResult<Boolean>

    public fun blacklisted(item: ItemDescriptor): Boolean

    public fun maxStackSize(item: ItemDescriptor): PlatformResult<Int>

    public fun give(
        player: CellPlayer,
        item: ItemDescriptor
    ): PlatformResult<ItemGrantResult>

    public fun count(
        player: CellPlayer,
        item: ItemDescriptor
    ): PlatformResult<Int>

    public fun take(
        player: CellPlayer,
        item: ItemDescriptor
    ): PlatformResult<Void>

    public fun heldItemId(player: CellPlayer): PlatformResult<String>

}
